framework/src/materials/VectorMagnitudeFunctorMaterial.C

```diff
@@ -20,38 +20,84 @@ VectorMagnitudeFunctorMaterialTempl<is_ad>::validParams()
   InputParameters params = FunctorMaterial::validParams();
   params += SetupInterface::validParams();
   params.set<ExecFlagEnum>("execute_on") = {EXEC_ALWAYS};
-  params.addClassDescription("This class takes up to three functors corresponding to vector "
-                             "components and computes the Euclidean norm from them.");
-  params.addRequiredParam<MooseFunctorName>("x_functor",
-                                            "The functor corresponding to the x component.");
+  params.addClassDescription(
+      "This class takes up to three scalar-valued functors corresponding to vector "
+      "components *or* a single vector functor and computes the Euclidean norm.");
+  params.addParam<MooseFunctorName>("x_functor", "The functor corresponding to the x component.");
   params.addParam<MooseFunctorName>(
       "y_functor", 0, "The functor corresponding to the y component.");
   params.addParam<MooseFunctorName>(
       "z_functor", 0, "The functor corresponding to the z component.");
-  params.addRequiredParam<MooseFunctorName>("vector_functor_name",
-                                            "The name of the vector functor that we are creating.");
+  params.addRequiredParam<MooseFunctorName>(
+      "vector_magnitude_name", "The name of the vector magnitude functor that we are creating.");
+  params.addParam<MooseFunctorName>(
+      "vector_functor", "The name of a vector functor that we will take the magnitude of.");
   return params;
 }
 
 template <bool is_ad>
 VectorMagnitudeFunctorMaterialTempl<is_ad>::VectorMagnitudeFunctorMaterialTempl(
     const InputParameters & parameters)
   : FunctorMaterial(parameters),
-    _x(getFunctor<GenericReal<is_ad>>("x_functor")),
+    _x(isParamValid("x_functor") ? &getFunctor<GenericReal<is_ad>>("x_functor") : nullptr),
     _y(getFunctor<GenericReal<is_ad>>("y_functor")),
-    _z(getFunctor<GenericReal<is_ad>>("z_functor"))
+    _z(getFunctor<GenericReal<is_ad>>("z_functor")),
+    _vector_functor(isParamValid("vector_functor")
+                        ? &getFunctor<VectorValue<GenericReal<is_ad>>>("vector_functor")
+                        : nullptr)
 {
+  auto check_error =
+      [this, &parameters](const std::string & scalar_functor, const bool must_equal_one)
+  {
+    auto error_message = [&scalar_functor, this]()
+    {
+      mooseError("Either a '",
+                 scalar_functor,
+                 "' or 'vector_functor' parameter "
+                 "must be provided to '",
+                 name(),
+                 "'");
+    };
+    const unsigned short sum =
+        parameters.isParamSetByUser(scalar_functor) + parameters.isParamSetByUser("vector_functor");
+    if (must_equal_one)
+    {
+      if (sum != 1)
+        error_message();
+    }
+    else
+    {
+      if (sum > 1)
+        error_message();
+    }
+  };
+
+  check_error("x_functor", true);
+  check_error("y_functor", false);
```

Comment:
why not must_equal one?
if there's y and a vector we should error?

Reply:
equaling zero is fine for the y and z checks, e.g. it's ok if a user provides just an `x_functor`. If `must_equal_one` is false, then we should for whether the sum is `> 1` so providing *both* is an error but providing zero is not.

Reply:
oh ok you merged that first check in
sounds good